### PR TITLE
refactor(runner): cut api start-time writer to canonical alias

### LINKED ISSUE #29998: refactor(runner): cut API start-time writer to canonical alias
Parent EPIC: #28914

## Background

Verified against clean current `main@528df1c9581610cc43644744abcbf1b89e06f304` on 2026-08-28.

`VM0_API_START_TIME` is trusted Runner-to-Guest bootstrap metadata. It is never part of the untrusted user/model/connector environment: `build_env_json_for_run` constructs bootstrap state separately from `ExecutionContext.environment`, and the canonical alias is already registered in the fail-closed runner-owned key contract.

The dual reader and same-checkout test cleanup are complete:

- Reader issue #28982 and PR #29022 merged as `928d53b17819c1c82f76da3aa8e4e672c69431d1`. Guest Agent accepts absent, legacy-only, canonical-only, and equal-dual API start time input; conflicting dual values fail closed without values; existing empty and non-Unicode behavior is preserved.
- The first published Runner release containing that reader is `runner-rs-v0.171.9` at exact target `feda3aafb652aca6dd3369ef18dd802635349e7f` on 2026-08-24.
- Test-writer issue #29953 and PR #29957 merged as `6b1f6b328efc2a747fbac17abafae66c70379d19`; all four ordinary same-checkout Guest Agent API start-time writers now use `OKOU_API_START_TIME` while the production Runner writer remains legacy-only.
- Post-sink telemetry PR #29927 merged as `784a12fd385e00a2f28306873c7ebdb747013410` and is deployed in production release `da100a97131334402bc194283be13feafa796e4b`.

Production release `528df1c9581610cc43644744abcbf1b89e06f304` completed successfully at https://github.com/vm0-ai/vm0/actions/runs/33132216461, including both Runner architecture promotions. Rollback dashboard #6792 currently retains exactly `528df1c9581610cc43644744abcbf1b89e06f304`, `da100a97131334402bc194283be13feafa796e4b`, `b095c5abb22e1d1dd5ab8194ccb0eae9ae5affdc`, `f8145c9b4ec150fee574c31155d42a09069c0ab3`, `4855fb80e0cce894dce180ab0594845c9c3295e2`, `edebf90da51f6e86a95002499be2f6b4b4e679e6`, and `ef41199bc788ed8ee6b192049389565c1a101c81`. GitHub comparison reports every target ahead of the reader merge by 467–549 commits, behind by zero, with merge base exactly `928d53b17819c1c82f76da3aa8e4e672c69431d1`. The reader floor has been deployed far longer than the documented two-hour Guest runtime plus bounded finalization/reuse window, and no supported rollback target predates it.

A fixed, value-free Axiom query over `2026-08-28T01:03:00Z..2026-08-28T01:33:00Z` was terminal and non-partial. It examined 124,731 `vm0-sandbox-telemetry-system-prod` rows and found 85 exact `run_metadata_env_source key=OKOU_API_START_TIME` markers across 85 distinct runs, from `2026-08-28T01:03:08.498Z` through `2026-08-28T01:32:33.337Z`. All 85 were `legacy-only`; canonical-only, dual, and conflict were zero. No timestamp value was selected or returned.

Current main has exactly one production writer:

- `crates/runner/src/executor/env.rs:605-611` inserts `guest_contracts::env::API_START_TIME_ENV`, serializing the optional API timestamp unchanged and emitting an empty string when it is unavailable.

Focused Runner coverage is at:

- `crates/runner/src/executor/tests/env_tests.rs:438-490` — the required-key scenario currently requires the legacy key with an empty value and rejects every canonical run-metadata key.
- `crates/runner/src/executor/tests/env_tests.rs:1400-1407` — the explicit timestamp scenario currently requires the exact millisecond value under the raw legacy literal.

Shared contract documentation is at `crates/guest-contracts/src/env.rs:105-112`. A complete current-main search found 26 API start-time literal/symbol references across 13 files; only the one Runner insertion is a production writer. The dedicated Guest reader, source telemetry, compatibility matrix, timing diagnostic, cleanup lists, and constant assertions must remain.

The fully paginated open-PR audit covered 30 open PRs and all 559/559 declared changed files with zero mismatch. Two PRs touch broad scoped files but neither overlaps the writer/test/doc semantics:

- #29943 touches private guest-file batching in `env.rs` and deletes/replaces private-file tests in `env_tests.rs`; it does not touch `build_env_json_with_host_env_inner`, either API start-time assertion, or the shared constants.
- stale #25722 adds Cloudflare Access constants/host injection and related assertions; it does not touch API start-time keys, values, readers, or tests.

These are inventory, not functional dependencies or ordering blockers. Refresh current main, the complete caller inventory, and every open-PR changed-file page immediately before editing. Normal first-merge precedence applies.

## Design

Switch only the production Runner API start-time writer to `guest_contracts::env::CANONICAL_API_START_TIME_ENV`.

Preserve the value contract byte-for-byte:

- when `context.api_start_time` is present, serialize the same integer Unix epoch milliseconds with `to_string()`;
- when absent, continue emitting an empty string;
- emit exactly one canonical entry and no legacy entry.

Update the two focused Runner assertions so they independently prove canonical empty-value behavior, canonical exact-value behavior, and legacy absence. In the required-key test, remove only `CANONICAL_API_START_TIME_ENV` from the Stage-1 canonical-absence loop; the other three run-metadata writers remain legacy-only and their existing negative assertions must stay unchanged.

Update only the adjacent shared rustdoc to record that Runner writes the canonical alias while Guest readers retain `API_START_TIME_ENV` as the rollback fallback.

Do not dual-write. Every active and rollback-retained reader accepts canonical-only input, while dual writing would retain the legacy surface and introduce an unnecessary conflict state.

The key trap is family separation: this issue moves only API start time. `VM0_SANDBOX_ID`, `VM0_SANDBOX_REUSE_RESULT`, and `VM0_WORKSPACE_REUSE_RESULT` share the same builder and reader family but remain separate production writer cutovers. Do not change them merely because they are adjacent or share one Stage-1 test loop.

## Explicit non-goals

- Do not change the other three run-metadata writers, resume-session writer, execution-timeout writer, API URL/token writers, private payload writers, tuning writers, CLI agent type, or any unrelated bootstrap key.
- Do not change Guest Agent's dual reader, source event inventory/emission, conflict/empty/non-Unicode behavior, timing parsing/diagnostics, E2E computation, or compatibility matrix.
- Do not remove or rename `API_START_TIME_ENV`, change the runner-owned ownership predicate, or remove any legacy reader/constant/test.
- Do not change user environment separation, merge precedence, value serialization, optionality, time units, clock source, run scheduling, sandbox reuse, checkpoint timing, reporting, or CLI child exposure.
- Do not absorb #29943 or #25722, modify/rebase/push to another contributor's PR, or wait for either PR merely to avoid conflicts.
- Do not perform a production release, deployment, sandbox drain, reader/telemetry removal, namespace-guard removal, or final cleanup.

## Rollout and rollback

- New Runner writer plus every active or rollback-retained Guest reader: canonical-only API start time is accepted with unchanged empty/present semantics.
- Rolled-back Runner writer plus every active or rollback-retained Guest reader: legacy-only API start time remains accepted.
- The legacy constant, reader, value-free source telemetry, and compatibility tests remain deployed.
- Reader removal is a separate later issue gated on an exact production release of this writer, a bounded canonical-only observation window, expiry of supported rollback writers, and zero legacy-only/dual/conflict observations.

## Acceptance criteria

- [ ] The sole production Runner writer emits exactly one `OKOU_API_START_TIME` entry and no `VM0_API_START_TIME` entry.
- [ ] Present timestamps retain the exact decimal Unix epoch millisecond value; absent timestamps retain the exact empty-string value.
- [ ] The required-key test proves canonical empty-value presence and legacy absence while leaving the three adjacent Stage-1 run-metadata writers legacy-only.
- [ ] The explicit timestamp test proves the exact canonical value and legacy absence.
- [ ] Guest reader, source telemetry, timing diagnostics/semantics, compatibility matrix, ownership contract, and all unrelated writers are byte-identical.
- [ ] Shared rustdoc accurately records the canonical-only Runner writer and retained legacy reader fallback.
- [ ] The diff is limited to `crates/runner/src/executor/env.rs`, `crates/runner/src/executor/tests/env_tests.rs`, and `crates/guest-contracts/src/env.rs`, with no unrelated cleanup or absorption of #29943/#25722.
- [ ] Rust formatting, strict Runner/Guest Contracts Clippy and check, warnings-denied docs, and the focused Runner env tests pass where locally supported; protected PR CI is authoritative for linked tests constrained by local memory.
- [ ] The PR references #28914 and this issue, records the refreshed base/head, fixed Axiom window, rollback ancestry, and fully paginated overlap audit, and merges through normal protected review and queue gates.

## Delivery

Create one focused PR from a fresh clean current main. Use a conventional lowercase commit and PR subject. Do not run the full local test suite and do not start a dev server. After opening the explicit PR, invoke `/pr-auto <PR URL>` in the same implementation thread and continue until GitHub reports `MERGED` or a genuine blocker is recorded. Do not use admin merge or bypass branch protection.

### PR DESCRIPTION
**Pull request by lancy**
Closes #29998.
Parent EPIC: #28914.

## Summary

- switch only `build_env_json_with_host_env_inner` API start-time output from `API_START_TIME_ENV` to `CANONICAL_API_START_TIME_ENV`
- preserve present decimal Unix epoch milliseconds via the unchanged `to_string()` path and preserve absent timestamps as the unchanged empty string
- prove canonical empty/present values and legacy absence while leaving sandbox ID, sandbox reuse, and workspace reuse writers legacy-only
- update only the adjacent shared rustdoc to describe the canonical Runner writer and retained legacy Guest-reader rollback fallback

The deployed Guest dual reader, source telemetry, conflict/empty/non-Unicode behavior, timing diagnostics and computation, ownership predicate, user-environment separation, serialization, clock source, sandbox lifecycle, reporting, CLI exposure, and every unrelated writer remain unchanged. This does not dual-write and does not remove a reader, constant, test, telemetry item, or namespace guard.

## Rollout evidence

Reader precedent #28982 / #29022 merged as `928d53b17819c1c82f76da3aa8e4e672c69431d1` and first shipped in Runner `v0.171.9` at `feda3aafb652aca6dd3369ef18dd802635349e7f`. Same-checkout cleanup #29957 merged as `6b1f6b328efc2a747fbac17abafae66c70379d19`; post-sink telemetry #29927 merged as `784a12fd385e00a2f28306873c7ebdb747013410` and is deployed in `da100a97131334402bc194283be13feafa796e4b`.

The fixed, terminal, non-partial Axiom window was `2026-08-28T01:03:00Z..2026-08-28T01:33:00Z`: 124,731 system-telemetry rows, 85 exact API-start markers across 85 distinct runs, from `2026-08-28T01:03:08.498Z` through `2026-08-28T01:32:33.337Z`. All 85 were legacy-only; canonical-only, dual, and conflict were zero. No timestamp value was selected or returned.

Every retained rollback target is ahead of the reader merge, behind by zero, and has merge base exactly `928d53b17819c1c82f76da3aa8e4e672c69431d1`:

| Rollback target | Ahead | Behind |
| --- | ---: | ---: |
| `528df1c9581610cc43644744abcbf1b89e06f304` | 549 | 0 |
| `da100a97131334402bc194283be13feafa796e4b` | 521 | 0 |
| `b095c5abb22e1d1dd5ab8194ccb0eae9ae5affdc` | 512 | 0 |
| `f8145c9b4ec150fee574c31155d42a09069c0ab3` | 504 | 0 |
| `4855fb80e0cce894dce180ab0594845c9c3295e2` | 491 | 0 |
| `edebf90da51f6e86a95002499be2f6b4b4e679e6` | 477 | 0 |
| `ef41199bc788ed8ee6b192049389565c1a101c81` | 467 | 0 |

## Base, head, and complete inventory

- clean pre-edit implementation base: `076afcef21eda68013427bb5636c971d506c992c`
- publication-time `origin/main`: `97414e6c34b2241df1cbcf87fa85fa6248cf41d6`
- published head: `643eedaf0d46b727fc6a5bd6cdeb415c4e67aa88`
- pre-edit search `API_START_TIME_ENV|OKOU_API_START_TIME|VM0_API_START_TIME`: exactly 26 references across 13 files; the sole production writer was `crates/runner/src/executor/env.rs:606`
- post-edit search: 27 references across the same 13 files because the explicit value test now adds a focused legacy-absence assertion; reader, telemetry, timing, compatibility, cleanup, constant, and ownership references remain

The fully paginated pre-edit audit covered 32 open PRs and all 579 declared / 579 fetched changed files, with zero mismatches. Exact scoped overlap hunks were:

- #29943, `crates/runner/src/executor/env.rs`: `@@ -9,7 +9,7 @@` and `@@ -480,34 +480,49 @@`; private-file batching imports and `write_required_agent_files`, not `build_env_json_with_host_env_inner` or API start time.
- #29943, `crates/runner/src/executor/tests/env_tests.rs`: `@@ -13,15 +13,14 @@`, `@@ -1175,97 +1174,6 @@`, and `@@ -1343,53 +1251,6 @@`; private-file test imports/removals, not either API start-time assertion.
- stale #25722, `crates/guest-contracts/src/env.rs`: `@@ -60,6 +60,12 @@` and `@@ -377,6 +383,8 @@`; Cloudflare Access constants and ownership entries.
- stale #25722, `crates/runner/src/executor/env.rs`: `@@ -472,6 +472,16 @@` and `@@ -667,14 +677,35 @@`; Cloudflare Access injection and host configuration.
- stale #25722, `crates/runner/src/executor/tests/env_tests.rs`: `@@ -10,10 +10,10 @@`, `@@ -571,6 +571,14 @@`, `@@ -629,6 +637,8 @@`, `@@ -680,6 +690,10 @@`, and `@@ -696,6 +710,8 @@`; Cloudflare Access imports and assertions.

#29943 merged normally during local validation as `97414e6c34b2241df1cbcf87fa85fa6248cf41d6`; this branch neither modified nor absorbed it. The pre-publication audit then covered 30 open PRs and all 565 declared / 565 fetched files with zero mismatches; stale #25722 was the sole open scoped overlap. Normal first-merge precedence applies.

## Verification

Passed locally on Rust 1.98 with one build job:

- `cargo fmt --manifest-path crates/Cargo.toml --all`
- `cargo fmt --manifest-path crates/Cargo.toml --all -- --check`
- `cargo clippy --manifest-path crates/Cargo.toml -p runner -p guest-contracts --all-targets --all-features -- -D warnings`
- `cargo check --manifest-path crates/Cargo.toml -p runner -p guest-contracts --all-targets --all-features`
- `RUSTDOCFLAGS='-D warnings' cargo doc --manifest-path crates/Cargo.toml -p runner -p guest-contracts --all-features --no-deps`
- `cargo test --manifest-path crates/Cargo.toml -p guest-contracts` (134 tests passed across unit and integration targets; doc tests passed)
- `git diff --check`

The exact Runner test command for `executor::tests::env_tests::build_env_json_required_keys` compiled through the Runner test target and entered the silent final-link phase, then the local 3.8 GiB/no-swap environment killed it with exit 137. No compiler diagnostic or test assertion ran. The second exact test would require the same unavailable linked binary, so it was not retried. Strict all-targets check passed; protected PR CI is authoritative for linked Runner execution.

## Fallbacks

- `crates/guest-agent/src/env.rs:run_metadata_env_or_empty` retains `API_START_TIME_ENV` as the rollback fallback for a rolled-back Runner writer. Remove it only in the later reader-removal issue after an exact writer release, canonical-only observation window, supported rollback-writer expiry, and zero legacy-only/dual/conflict evidence.
- Reverting this focused PR restores the legacy-only Runner writer without changing any reader, value semantics, configuration, or persisted state.

### CODE REVIEW

**Comment by lancy** [comment]
LGTM


PR_AUTO_REVIEW_COMMIT: 643eedaf0d46b727fc6a5bd6cdeb415c4e67aa88
PR_AUTO_REVIEW_VERDICT: LGTM

### Summary

This focused refactor switches only the production Runner API start-time key to the existing canonical alias, preserves the exact present decimal-millisecond and absent empty-string values, and updates the two focused assertions plus adjacent shared rustdoc. The complete three-file diff retains the deployed dual reader, value-free source telemetry, legacy rollback fallback, timing behavior, ownership protection, and all adjacent run-metadata writers.

### Review Criteria

| Source | Core item | Result | Evidence |
| --- | --- | --- | --- |
| `REVIEW.md` | Correctness and edge cases | Pass | `crates/runner/src/executor/env.rs` changes only the map key; the unchanged `to_string()` and `unwrap_or_default()` preserve present and absent values, while both focused tests prove canonical presence and legacy absence. |
| `REVIEW.md` | Security and type/interface safety | Pass | The canonical key was already in `EXPLICIT_RUNNER_OWNED_ENV_KEYS`; no user environment, secret, parser, type, API, or persisted interface changes. |
| `REVIEW.md` | Style and maintainability | Pass | The diff uses the existing shared contract constant, adds no abstraction or duplicated logic, and limits documentation to the adjacent constants. |
| `REVIEW.md` | Refactor coverage | Pass | `build_env_json_required_keys` covers canonical empty output and family separation; `build_env_json_with_api_start_time` covers the exact decimal value; both assert legacy absence. |
| `REVIEW.md` | DB/JSONB and feature-regression gates | N/A | No database, JSONB, API, UI, feature switch, or user-facing flow changes. |
| `docs/bad-smell.md` | Fail fast and avoid suppressions or fabricated defaults | Pass | No error path, fallback branch, suppression, dynamic import, or configuration default is added; the pre-existing empty-value contract remains explicit. |
| `docs/bad-smell.md` | Document modified interfaces | Pass | `crates/guest-contracts/src/env.rs` now states that Runner writes the canonical alias and Guest readers retain the legacy alias for rollback. |
| `docs/testing.md` | Test behavior at the established integration boundary | Pass | Existing Runner executor tests exercise the real environment builder and assert emitted map behavior rather than mocks or internal call counts. |
| `docs/testing.md` | Preserve reachable rollout combinations | Pass | Canonical-only new Runner input is accepted by every active and retained reader; rolled-back legacy-only writers remain accepted by the unchanged dual reader. |
| `docs/testing/rust-testing.md` | Rust test placement and process-environment safety | Pass | The changed synchronous Runner tests stay in the established executor test module, mutate no process environment, and require no mocks or external infrastructure. |
| `docs/fallback.md` | Keep only bounded, reachable rollout compatibility | Pass | The diff introduces no fallback; it retains the already-deployed legacy Guest reader for rolled-back Runner writers, with #28914/#29998 production and rollback gates recorded for later removal. |
| `docs/deployment-compatibility.md` | Prepare reader before writer and defer cleanup | Pass | Reader #29022 and telemetry #29927 precede this writer cutover; rollback ancestry is complete, and no legacy reader/telemetry/constant is removed in the migration step. |

### Findings

#### Critical (P0)

- None.

#### High Priority (P1)

- None.

### Testing

- Coverage: Adequate — absent and exact present writer values, canonical-only output, legacy absence, and separation from the three adjacent Stage-1 metadata writers are explicitly covered; the pre-existing Guest compatibility matrix remains unchanged.
- Conventions: Compliant — Rust formatting, strict Runner/Guest Contracts Clippy and all-targets check, warning-denied rustdoc, 134 Guest Contracts tests, and `git diff --check` passed locally.
- Local resource boundary: the exact Runner test target reached final link and was killed with exit 137 by the 3.8 GiB/no-swap host after all-targets check had type-checked it; protected CI is authoritative for linked execution.
